app/scripts/lib/ppom/ppom-util.ts

```diff
@@ -1,19 +1,180 @@
+  }
+}
+
+export function generateSecurityAlertId(): string {
+  return uuid();
+}
+
+export async function updateSecurityAlertResponse({
+  appStateController,
+  method,
+  securityAlertId,
+  securityAlertResponse,
+  signatureController,
+  transactionController,
+}: {
+  appStateController: AppStateController;
+  method: string;
+  securityAlertId: string;
+  securityAlertResponse: SecurityAlertResponse;
+  signatureController: SignatureController;
+  transactionController: TransactionController;
+}) {
+  const isSignatureRequest = SIGNING_METHODS.includes(method);
+
+  const confirmation = await findConfirmationBySecurityAlertId(
+    securityAlertId,
+    method,
+    signatureController,
+    transactionController,
+  );
+
+  if (isSignatureRequest) {
+    appStateController.addSignatureSecurityAlertResponse(securityAlertResponse);
+  } else {
+    transactionController.updateSecurityAlertResponse(
+      confirmation.id,
+      securityAlertResponse,
+    );
+  }
+}
+
+export function handlePPOMError(
+  error: unknown,
+  logMessage: string,
+): SecurityAlertResponse {
+  const errorData = getErrorData(error);
+  const description = getErrorMessage(error);
+
+  sentry?.captureException(error);
+  console.error(logMessage, errorData);
+
+  return {
+    ...SECURITY_ALERT_RESPONSE_ERROR,
+    description,
+  };
+}
+
+async function usePPOM(
+  request: JsonRpcRequest,
+  securityAlertId: string,
+  ppom: PPOM,
+): Promise<SecurityAlertResponse> {
+  try {
+    const normalizedRequest = normalizePPOMRequest(request);
+    const ppomResponse = await ppom.validateJsonRpc(normalizedRequest);
+
+    return {
+      ...ppomResponse,
+      securityAlertId,
+    };
+  } catch (error: unknown) {
+    return handlePPOMError(error, 'Error validating JSON RPC using PPOM: ');
+  }
+}
+
+function normalizePPOMRequest(request: JsonRpcRequest): JsonRpcRequest {
   if (request.method !== METHOD_SEND_TRANSACTION) {
     return request;
   }
 
-  const transactionParams = request.params?.[0] || {};
+  const transactionParams = (request.params?.[0] || {}) as TransactionParams;
   const normalizedParams = normalizeTransactionParams(transactionParams);
 
   return {
     ...request,
     params: [normalizedParams],
   };
 }
+
+function getErrorMessage(error: unknown) {
+  if (error instanceof Error) {
+    return `${error.name}: ${error.message}`;
+  }
+
+  return JSON.stringify(error);
+}
+
+function getErrorData(error: unknown) {
+  if (typeof error === 'object' || typeof error === 'string') {
+    return error;
+  }
+
+  return JSON.stringify(error);
+}
+
+async function findConfirmationBySecurityAlertId(
+  securityAlertId: string,
+  method: string,
+  signatureController: SignatureController,
+  transactionController: TransactionController,
+) {
+  const isSignatureRequest = SIGNING_METHODS.includes(method);
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    let confirmation;
```

Comment:
Not related to this PR, but it will be nice if we can get rid of this wait.

Reply:
Totally agree, out of scope now, but a nice alternative could be just adding controller specific state listeners.